cpp/src/parquet/size_statistics.h

```diff
@@ -0,0 +1,156 @@
+  /// written)`. To determine unencoded sizes of other types readers can use
+  /// schema information multiplied by the number of non-null and null values.
+  /// The number of null/non-null values can be inferred from the histograms
+  /// below.
+  ///
+  /// For example, if a column chunk is dictionary-encoded with dictionary
+  /// ["a", "bc", "cde"], and a data page contains the indices [0, 0, 1, 2],
+  /// then this value for that data page should be 7 (1 + 1 + 2 + 3).
+  ///
+  /// This field should only be set for types that use BYTE_ARRAY as their
+  /// physical type.
+  ///
+  /// \returns unencoded and uncompressed byte size of the BYTE_ARRAY column,
+  /// or std::nullopt for other types.
+  std::optional<int64_t> unencoded_byte_array_data_bytes() const;
+
+  /// \brief Merge two SizeStatistics of the same column.
+  ///
+  /// It is used to merge size statistics from all pages of the same column chunk.
+  void Merge(const SizeStatistics& other);
+
+ private:
+  friend class SizeStatisticsBuilder;
+  SizeStatistics(const void* size_statistics, const ColumnDescriptor* descr);
+
+  // PIMPL Idiom
+  SizeStatistics();
+  class SizeStatisticsImpl;
+  std::unique_ptr<SizeStatisticsImpl> impl_;
+};
+
+/// \brief Builder to create a SizeStatistics.
+class PARQUET_EXPORT SizeStatisticsBuilder {
+ public:
+  /// \brief API convenience to get a SizeStatisticsBuilder.
+  static std::unique_ptr<SizeStatisticsBuilder> Make(const ColumnDescriptor* descr);
+
+  ~SizeStatisticsBuilder();
+
+  /// \brief Add repetition levels to the histogram.
+  /// \param num_levels number of repetition levels to add.
+  /// \param rep_levels repetition levels to add.
+  void WriteRepetitionLevels(int64_t num_levels, const int16_t* rep_levels);
+
+  /// \brief Add definition levels to the histogram.
+  /// \param num_levels number of definition levels to add.
+  /// \param def_levels definition levels to add.
+  void WriteDefinitionLevels(int64_t num_levels, const int16_t* def_levels);
+
+  /// \brief Add repeated repetition level to the histogram.
+  /// \param num_levels number of repetition levels to add.
+  /// \param rep_level repeated repetition level value.
+  void WriteRepetitionLevel(int64_t num_levels, int16_t rep_level);
+
+  /// \brief Add repeated definition level to the histogram.
+  /// \param num_levels number of definition levels to add.
+  /// \param def_level repeated definition level value.
+  void WriteDefinitionLevel(int64_t num_levels, int16_t def_level);
+
+  /// \brief Add spaced BYTE_ARRAY values.
+  /// \param[in] values pointer to values of BYTE_ARRAY type.
+  /// \param[in] valid_bits pointer to bitmap representing if values are non-null.
+  /// \param[in] valid_bits_offset offset into valid_bits where the slice of data begins.
+  /// \param[in] num_spaced_values length of values in values/valid_bits to inspect.
+  void WriteValuesSpaced(const ByteArray* values, const uint8_t* valid_bits,
```

Comment:
as commented above, I wonder if it is possible to not interwrine values spaced (and the Array option below) into this interface.

Reply:
The logic is anyway required. Perhaps we can provide only the dense interface here and move the logic of dealing with nulls & arrow array to the caller?